src/common/servers/serverManager.ts

```diff
@@ -0,0 +1,426 @@
+            });
+        return selectedTab;
+    }
+
+    updateServerOrder = (serverOrder: string[]) => {
+        log.debug('updateServerOrder', serverOrder);
+
+        this.serverOrder = serverOrder;
+        this.persistServers();
+    }
+
+    updateTabOrder = (serverId: string, tabOrder: string[]) => {
+        log.debug('ServerManager.updateTabOrder', serverId, tabOrder);
+
+        this.tabOrder.set(serverId, tabOrder);
+        this.persistServers();
+    }
+
+    addServer = (server: Team) => {
+        const newServer = new MattermostServer(server, false);
+        this.servers.set(newServer.id, newServer);
+        this.serverOrder.push(newServer.id);
+        const tabOrder: string[] = [];
+        getDefaultTabs().forEach((tab) => {
+            const newTab = this.getTabView(newServer, tab.name, tab.isOpen);
+            this.tabs.set(newTab.id, newTab);
+            tabOrder.push(newTab.id);
+        });
+        this.tabOrder.set(newServer.id, tabOrder);
+
+        this.emit(SERVERS_URL_MODIFIED, [newServer.id]);
+        this.persistServers();
+        return newServer;
+    }
+
+    editServer = (serverId: string, server: Team) => {
+        const existingServer = this.servers.get(serverId);
+        if (!existingServer) {
+            return;
+        }
+
+        let urlModified;
+        if (existingServer.url.toString() !== urlUtils.parseURL(server.url)?.toString()) {
+            urlModified = () => this.emit(SERVERS_URL_MODIFIED, [serverId]);
+        }
+        existingServer.name = server.name;
+        existingServer.updateURL(server.url);
+        this.servers.set(serverId, existingServer);
+
+        this.tabOrder.get(serverId)?.forEach((tabId) => {
+            const tab = this.tabs.get(tabId);
+            if (tab) {
+                tab.server = existingServer;
+                this.tabs.set(tabId, tab);
+            }
+        });
+
+        urlModified?.();
+        this.persistServers();
+    }
+
+    removeServer = (serverId: string) => {
+        this.tabOrder.get(serverId)?.forEach((tabId) => this.tabs.delete(tabId));
+        this.tabOrder.delete(serverId);
+        this.lastActiveTab.delete(serverId);
+
+        const index = this.serverOrder.findIndex((id) => id === serverId);
+        this.serverOrder.splice(index, 1);
+        this.remoteInfo.delete(serverId);
+        this.servers.delete(serverId);
+
+        this.persistServers();
+    }
+
+    toggleTab = (tabId: string, isOpen: boolean) => {
+        const tab = this.tabs.get(tabId);
+        if (!tab) {
+            return;
+        }
+        tab.isOpen = isOpen;
+
+        this.persistServers();
+    }
+
+    updateLastActive = (tabId: string) => {
+        const tab = this.tabs.get(tabId);
+        if (!tab) {
+            return;
+        }
+        this.lastActiveTab.set(tab.server.id, tabId);
+
+        this.currentServerId = tab.server.id;
+
+        const serverOrder = this.serverOrder.findIndex((srv) => srv === tab.server.id);
```

Comment:
remember that `findIndex` could return `-1`

Reply:
If this ever happens we should be hard crashing, but I can be explicit about the error thrown.